src/app/server-cluster/ServerClusterInterface.h

```diff
@@ -0,0 +1,223 @@
+    IntrusiveSingleLinkedList() : mNext(static_cast<SELF *>(this)) {}
+
+    // IMPLEMENTATION DETAILS:
+    //   Since `mNext == this` is used as a marker for "is in a list",
+    //   the assignment of these interfaces is overloaded even for the move operator.
+    IntrusiveSingleLinkedList(IntrusiveSingleLinkedList && other) :
+        mNext((other.mNext == &other) ? static_cast<SELF *>(this) : other.mNext)
+    {}
+    IntrusiveSingleLinkedList(const IntrusiveSingleLinkedList & other) :
+        mNext((other.mNext == &other) ? static_cast<SELF *>(this) : other.mNext)
+    {}
+    IntrusiveSingleLinkedList & operator=(const IntrusiveSingleLinkedList & other)
+    {
+        if (&other != this)
+        {
+            mNext = (other.mNext == &other) ? static_cast<SELF *>(this) : other.mNext;
+        }
+        return *this;
+    }
+    IntrusiveSingleLinkedList & operator=(const IntrusiveSingleLinkedList && other)
+    {
+        mNext = (other.mNext == &other) ? static_cast<SELF *>(this) : other.mNext;
+        return *this;
+    }
+
+    // Internal details for processing
+
+    /// This is NOT a STABLE API.
+    ///
+    /// Determines if this object is part of a linked list already or not.
+    [[nodiscard]] bool IsInList() const { return (mNext != this); }
+
+    /// This is NOT a STABLE API.
+    ///
+    /// Marks this object as not being part of a linked list
+    void SetNotInList() { mNext = static_cast<SELF *>(this); }
+
+    /// This is NOT a STABLE API. It allows interfaces to be stored within a registry, however
+    /// usage of this method MAY be changed in the future.
+    ///
+    /// Returns a "next" pointer when the ServerClusterInterface is assumed to be
+    /// part of a SINGLE linked list.
+    [[nodiscard]] SELF * GetNextListItem() const
+    {
+        VerifyOrDie(mNext != this);
+        return mNext;
+    }
+
+    /// This is NOT a STABLE API. It allows interfaces to be stored within a registry, however
+    /// usage of this method MAY be changed in the future.
+    ///
+    /// Sets the "next" pointer when the SELF is assumed to be
+    /// part of a SINGLE linked list.
+    ///
+    /// Returns the old value of "next"
+    SELF * SetNextListItem(SELF * value)
+    {
+        VerifyOrDie(value != this);
+        auto previousValue = mNext;
+        mNext              = value;
+        return previousValue;
+    }
+
+private:
+    // The mNext pointer has 2 (!) states:
+    //  - `this` means this item is NOT part of a linked list (used since we would generally
+    //    not allow loops)
+    //  - OTHER values, including nullptr, when this is part of a REAL LIST
+    SELF * mNext; /* = this (in constructor) */
+};
+
+} // namespace detail
+
+/// Defines an active cluster on an endpoint.
+///
+/// Provides metadata as well as interaction processing (attribute read/write and command handling).
+///
+/// Implementation note:
+///   - this class is highly coupled with ServerClusterInterfaceRegistry. The fact that it
```

Comment:
If ServerClusterInterfaceRegistry is optional ember compatibility code, could the state associated with managing these lists also be optional?

Reply:
I could not figure out a good way to do this and still highly optmize for RAM.

I think if we do decide to allow for another 100 or so bytes of RAM for tiny apps, we could make this a real map to pointers (we would need endpointid + pointer, so about 8 bytes for every mapped cluster ... a tiny thread sensor with 1 endpoint should have 10 root node clusters and 3 clusters on a endpoint like "contact sensor" or "light sensor") .. so every increase is multiplied by 13 or more as cluster counts increase.

Would be interested in suggestions, although at this time I would prefer them as followups.